edb/edgeql-parser/src/parser/mod.rs

```diff
@@ -394,6 +442,43 @@ impl<'s> Parser<'s> {
         Some(final_node.value)
     }
 
+    /// Lightweight version of act that checks if a token *could* be applied.
+    /// Returns next state.
+    fn can_act(&self, ctx: &'s Context, token: &Kind) -> Option<usize> {
+        let mut state = self.stack_top.state;
+
+        let mut node = &self.stack_top;
+        let mut pushed = 0;
+
+        loop {
+            // find next action
+            let action = ctx.spec.actions[state].get(token)?;
+
+            match action {
+                Action::Shift(next) => {
+                    return Some(*next);
+                }
+                Action::Reduce(reduce) => {
+                    // simulate reduce stack pops
+                    let mut to_pop = reduce.cnt;
+                    let cancel_out = usize::min(pushed, to_pop);
+                    to_pop -= cancel_out;
+                    pushed -= cancel_out;
+                    for _ in 0..to_pop {
+                        node = node.parent.as_ref().unwrap();
+                    }
+
+                    // get state of current stack top
+                    let stack_state = if reduce.cnt == 0 { state } else { node.state };
```

Comment:
Can't we always take `node.state`?

Reply:
No, because when we should have pushed onto the stack, we really didn't because I try to be conservative with allocations.

Instead, I save that into `pushed` count. We never really need the top stack node, because we pop nodes down from the stack at the beginning of the reduce. Unless if we don't because `reduce.cnt` is zero. In that case we have to look for the state the top node would have been, if we had pushed, and that is just state.

Complicated, but we don't need to know how this works because we have the proper `act` impl for that.

Reply:
I've added comments and named these "nodes that we should have added" as "ghost nodes"